lib-rt/CPy.h

```diff
@@ -69,24 +70,117 @@ static inline void CPy_FixupTraitVtable(CPyVTableItem *vtable, int count) {
+    // its __new__ function. Unfortunately there doesn't seem to be a
+    // good way to mix a C extension class and creating it via a
+    // metaclass. We need to do it anyways, though, in order to
+    // support subclassing Generic[T] prior to Python 3.7.
+    //
+    // We solve this with a kind of atrocious hack: create a parallel
+    // class using the metaclass, determine the bases of the real
+    // class by pulling them out of the parallel class, creating the
+    // real class, and then merging its dict back into the original
+    // class. There are lots of cases where this won't really work,
+    // but for the case of GenericMeta setting a bunch of properties
+    // on the class we should be fine.
+    if (metaclass != &PyType_Type) {
+        PyObject *ns = PyDict_New();
+        if (!ns) goto error;
+
+        dummy_class = (PyTypeObject *)PyObject_CallFunctionObjArgs(
+            (PyObject *)metaclass, name, bases, ns, NULL);
+        Py_DECREF(ns);
+        if (!dummy_class)
+            goto error;
+
+        Py_DECREF(bases);
+        bases = dummy_class->tp_bases;
+        Py_INCREF(bases);
+    }
+
+    // Allocate the type and then copy the main stuff in.  We create
+    // it with the type metaclass and then set the real metaclass
+    // later, since otherwise there are some extra strictness checks
+    // about the mro we fail.
+    t = (PyHeapTypeObject*)PyType_GenericAlloc(&PyType_Type, 0);
+    if (!t)
+        goto error;
     memcpy((char *)t + sizeof(PyVarObject),
            (char *)template_ + sizeof(PyVarObject),
            sizeof(PyTypeObject) - sizeof(PyVarObject));
 
-    PyObject *name = PyUnicode_FromString(template_->tp_name);
+    if (bases != old_bases) {
+        if (PyObject_SetAttrString((PyObject *)t, "__orig_bases__", old_bases) < 0)
+            goto error;
+    }
+
     t->ht_name = name;
     Py_INCREF(name);
     t->ht_qualname = name;
+    Py_XINCREF(bases);
+    t->ht_type.tp_bases = bases;
 
     if (PyType_Ready((PyTypeObject *)t) < 0)
-        return NULL;
+        goto error;
+
+    Py_INCREF(metaclass);
+    Py_TYPE(t) = metaclass;
+
+    if (dummy_class) {
+        if (PyDict_Merge(t->ht_type.tp_dict, dummy_class->tp_dict, 0) != 0)
+            goto error;
+        // This is the *really* tasteless bit. GenericMeta's __new__
+        // in certain versions of typing sets _gorg to point back to
+        // the class. We need to override it to keep it from pointing
+        // to the proxy.
+        if (PyDict_SetItemString(t->ht_type.tp_dict, "_gorg", (PyObject *)t) < 0)
+            goto error;
+        Py_DECREF(dummy_class);
+    }
+
+    if (PyObject_SetAttrString((PyObject *)t, "__module__", modname) < 0)
+        goto error;
+
+    if (init_subclass((PyTypeObject *)t, NULL))
+        goto error;
+
+    Py_XDECREF(bases);
+    Py_XDECREF(old_bases);
 
-    PyObject_SetAttrString((PyObject *)t, "__module__", modname);
+    return (PyObject *)t;
```

Comment:
Yeah, but mypyc values all basically need to be `PyObject *` so there needs to be a cast here or at all the call sites. I picked here but it could easily be the other way too.